In [1]:
import pandas as pd

Question 1
Importing Data

In [2]:

import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

<H3>Question 2</H3>
Installing geocoder

In [13]:
pip install geocoder

     |████████████████████████████████| 98 kB 9.3 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [14]:

import geocoder

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data

<H3>Question 3</H3>
Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
from geopy.geocoders import Nominatim

In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [24]:
pip install folium

     |████████████████████████████████| 94 kB 5.5 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [25]:
import folium

In [26]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [28]:
CLIENT_ID = 'TB5AIGTXUTWDE0UDFJAOJGKOQ3Q3UNQRVYLKOD0GRWY2WQ5H' 
CLIENT_SECRET = 'LKOTMEWA5ZUTPYTXWNYZPZJSNLJDN2WW0A0XNODT5DTXCDSN'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: TB5AIGTXUTWDE0UDFJAOJGKOQ3Q3UNQRVYLKOD0GRWY2WQ5H
CLIENT_SECRET:LKOTMEWA5ZUTPYTXWNYZPZJSNLJDN2WW0A0XNODT5DTXCDSN


Next, we create a function to get all the venue categories in Toronto

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)


Collecting the venues in Toronto for each Neighbourhood

In [30]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [31]:

venues_in_toronto.shape

(1324, 5)

In [32]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1          TTC stop #8380               Bus Stop  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood

In [33]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656          TTC stop #8380   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1311               43.628841               -79.520999              Wingporium   
1312               43.628841               -79.520999        South St. Burger   
1313               43.628841               -79.520999               Dollarama   
1314               43.628841               -79.520999          Healthy Planet   
1315               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                  Bus Stop  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1311            Wings Joint  
1312           Burger Joint  
1313         Discount Store  
1314        Supplement Shop  
1315                 Bakery  

[412 rows x 5 columns]

Checking for the maximum venue categories

In [34]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Scarborough Village   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.379817   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.744734               -79.239476   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Wine Bar                          The National Club  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[238 rows x 4 columns]

<h2>One Hot encoding theOne Hot encoding the venue Categories venue Categories</h2>

In [35]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1319                  0        0                   0             0   
1320                  0        0                   0             0   
1321                  0        0                   0             0   
1322                  0        0                   0             0   
1323                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1319               0                0                 0                    0   
1320               0                0                 0                    0   
1321               0                0                 0                    0   
1322               0                0                 0                    0   
1323               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Vegetarian / Vegan Restaurant  \
0                0         0  ...                              0   
1                0         0  ...                              0   
2                0         0  ...                              0   
3                0         0  ...                              0   
4                0         0  ...                              0   
...            ...       ...  ...                            ...   
1319             0         0  ...                              0   
1320             0         0  ...                              0   
1321             0         0  ...                              0   
1322             0         0  ...                              0   
1323             0         0  ...                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1319                 0            0                      0                0   
1320                 0            0                      0                0   
1321                 0            0                      0                0   
1322                 0            0                      0                0   
1323                 0            0                      0                0   

      Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0            0          0            0              0            0  
1            0          0            0              0            0  
2            0          0            0              0            0  
3            0          0            0              0            0  
4            0          0            0              0            0  
...        ...        ..

Adding the neighbourhood to the encoded dataframe

In [36]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Vegetarian / Vegan Restaurant  \
0                    0             0  ...                              0   
1                    0             0  ...                              0   
2                    0             0  ...                              0   
3                    0             0  ...                              0   
4                    0             0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 239 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [37]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0             0.000000           0.0  ...   
1               0.0             0.000000           0.0  ...   
2               0.0             0.000000           0.0  ...   
3               0.0             0.000000           0.0  ...   
4               0.0             0.041667           0.0  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 239 columns]

Let's make a function to get the top most common venue categories

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [39]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [41]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East            Hobby Shop   

       2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot               Skating Rink   
1         Athletics & Sports           Coffee Shop                        Pub   
2                       Bank              Pharmacy  Middle Eastern Restaurant   
3         Chinese Restaurant                  Bank        Japanese Restaurant   
4             Sandwich Place           Coffee Shop         Italian Restaurant   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Dessert Shop  Ethiopian Restaurant     Electronics Store   
1                      Gym           Golf Course     Electronics Store   
2                    Diner         Shopping Mall        Sandwich Place   
3         Department Store  Ethiopian Restaurant     Electronics Store   
4  Comfort Food Restaurant             Juice Bar               Butcher   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant             Drugstore             Donut Shop  
1                    Drugstore            Donut Shop                Dog Run  
2                  Supermarket         Deli / Bodega       Sushi Restaurant  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4          Japanese Restaurant      Sushi Restaurant                   Café

In [42]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [43]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [44]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 4, 0,
       1, 0, 0, 0, 0, 0, 1], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [46]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0                  Park   
1  43.725882 -79.315572             0.0           Pizza Place   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0       Food & Drink Shop               Bus Stop           Yoga Studio   
1             Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                    Park                 Bakery               Theater   
3  Furniture / Home Store      Accessories Store   Arts & Crafts Store   
4        Sushi Restaurant                    Gym            Hobby Shop   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0      Department Store     Electronics Store  Eastern European Restaurant   
1   Distribution Center      Department Store                 Dessert Shop   
2        Breakfast Spot  Gym / Fitness Center                Historic Site   
3           Event Space    Miscellaneous Shop                  Coffee Shop   
4              Beer Bar         Smoothie Shop                Burrito Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Drugstore            Donut Shop                Dog Run  
1    Dim Sum Restaurant                 Diner         Discount Store  
2     French Restaurant        Farmers Market            Event Space  
3              Boutique         Women's Store  Vietnamese Restaurant  
4        Sandwich Place                  Café  Portuguese Restaurant

Drop all the NaN values to prevent data skew

In [47]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [48]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [49]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             0.0           Pizza Place   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
7          North York             0.0                   Gym   
..                ...             ...                   ...   
96   Downtown Toronto             0.0           Coffee Shop   
97   Downtown Toronto             0.0                  Café   
99   Downtown Toronto             0.0      Sushi Restaurant   
100      East Toronto             0.0    Light Rail Station   
102         Etobicoke             0.0         Grocery Store   

      2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
1               Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                      Park                 Bakery               Theater   
3    Furniture / Home Store      Accessories Store   Arts & Crafts Store   
4          Sushi Restaurant                    Gym            Hobby Shop   
7                Beer Store    Japanese Restaurant        Clothing Store   
..                      ...                    ...                   ...   
96       Italian Restaurant                   Café            Restaurant   
97              Coffee Shop             Restaurant                 Hotel   
99         Ramen Restaurant               Beer Bar          Burger Joint   
100           Garden Center             Comic Shop            Restaurant   
102                  Bakery      Convenience Store        Discount Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1     Distribution Center      Department Store          Dessert Shop   
2          Breakfast Spot  Gym / Fitness Center         Historic Site   
3             Event Space    Miscellaneous Shop           Coffee Shop   
4                Beer Bar         Smoothie Shop         Burrito Place   
7              Restaurant           Coffee Shop           Supermarket   
..                    ...                   ...                   ...   
96                 Bakery     Indian Restaurant            Playground   
97     Seafood Restaurant           Art Gallery                   Pub   
99    Martial Arts School        Ice Cream Shop     Indian Restaurant   
100               Butcher            Skate Park         Burrito Place   
102        Sandwich Place          Burger Joint           Social Club   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1      Dim Sum Restaurant                 Diner         Discount Store  
2       French Restaurant        Farmers Market            Event Space  
3                Boutique         Women's Store  Vietnamese Restaurant  
4          Sandwich Place                  Café  Portuguese Restaurant  
7               Bike Shop        Discount Store   Caribbean Restaurant  
..                    ...                   ...                    ...  
96             Beer Store               Butcher  General Entertainment  
97                 Bakery              Tea Room              Speakeasy  
99                   Café            Restaurant    Japanese Restaurant  
100               Brewery        Farmers Market   Fast Food Restaurant  
102       Supplement Shop         Tanning Salon   Fast Food Restaurant  

[87 rows x 12 columns]

Cluster 2

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             1.0                  Park   
21              York             1.0                  Park   
35         East York             1.0                  Park   
64              York             1.0                  Park   
66        North York             1.0                  Park   
85       Scarborough             1.0            Playground   
91  Downtown Toronto             1.0                  Park   
98         Etobicoke             1.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop              Bus Stop           Yoga Studio   
21         Women's Store                  Pool         Deli / Bodega   
35     Convenience Store          Intersection           Yoga Studio   
64           Yoga Studio         Deli / Bodega  Ethiopian Restaurant   
66     Convenience Store           Yoga Studio      Department Store   
85                  Park          Intersection         Deli / Bodega   
91            Playground                 Trail           Yoga Studio   
98                 River           Yoga Studio         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue  \
0       Department Store            Electronics Store   
21  Ethiopian Restaurant            Electronics Store   
35      Department Store         Ethiopian Restaurant   
64     Electronics Store  Eastern European Restaurant   
66  Ethiopian Restaurant            Electronics Store   
85     Electronics Store  Eastern European Restaurant   
91        Discount Store             Department Store   
98  Ethiopian Restaurant            Electronics Store   

          7th Most Common Venue        8th Most Common Venue  \
0   Eastern European Restaurant                    Drugstore   
21  Eastern European Restaurant                    Drugstore   
35            Electronics Store  Eastern European Restaurant   
64                    Drugstore                   Donut Shop   
66  Eastern European Restaurant                    Drugstore   
85                    Drugstore                   Donut Shop   
91                 Dessert Shop           Dim Sum Restaurant   
98  Eastern European Restaurant                    Drugstore   

   9th Most Common Venue 10th Most Common Venue  
0             Donut Shop                Dog Run  
21            Donut Shop                Dog Run  
35             Drugstore             Donut Shop  
64               Dog Run    Distribution Center  
66            Donut Shop                Dog Run  
85               Dog Run    Distribution Center  
91                 Diner                Dog Run  
98            Donut Shop                Dog Run

Cluster 3

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant      Department Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6           Event Space  Ethiopian Restaurant     Electronics Store   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Eastern European Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

Cluster 4

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0             Locksmith        Baseball Field   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57         Farmers Market           Event Space  Ethiopian Restaurant   
101           Yoga Studio           Event Space  Ethiopian Restaurant   

    6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
57      Electronics Store  Eastern European Restaurant             Drugstore   
101     Electronics Store  Eastern European Restaurant             Drugstore   

    9th Most Common Venue 10th Most Common Venue  
57             Donut Shop                Dog Run  
101            Donut Shop                Dog Run

Cluster 5

In [54]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11        Farmers Market           Event Space  Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
11     Electronics Store  Eastern European Restaurant             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
11            Donut Shop                Dog Run